src/mawpy/steps/address_oscillation.py

```diff
@@ -0,0 +1,254 @@
+import logging
+
+import numpy as np
+import pandas as pd
+
+from mawpy.constants import USER_ID, UNIX_START_T, STAY_DUR, STAY_LAT, ORIG_LAT, ORIG_LONG, STAY_LONG, AO_COLUMNS
+from mawpy.utilities.preprocessing import get_preprocessed_dataframe, get_list_of_chunks_by_column, execute_parallel
+
+logger = logging.getLogger(__name__)
+
+
+def _transform_trace(row: pd.Series, df_columns: list) -> pd.Series:
+    """
+    Transforms a row of the dataframe by determining the duration, latitude, longitude,
+    and whether it's a stay point or not.
+
+    Args:
+        row (pd.Series): A row from the dataframe.
+
+    Returns:
+        pd.Series: Transformed row with duration, latitude, longitude, and stay/not stay indicator.
+    """
+    # Determine duration, defaulting to 1 if not present or stay_dur
+    duration = 1 if STAY_DUR not in df_columns or row[STAY_DUR] == -1 else row[STAY_DUR]
+
+    # Determine latitude, defaulting to original if not present or stay_lat
+    lat = row[ORIG_LAT] if STAY_LAT not in df_columns or row[STAY_LAT] == -1 else row[STAY_LAT]
+
+    # Determine longitude, defaulting to original if not present or stay_long
+    long = row[ORIG_LONG] if STAY_LONG not in df_columns or row[STAY_LONG] == -1 else row[STAY_LONG]
+
+    # Determine if it's a stay point, defaulting to 0 if not present or 1
+    stay_not = 0 if STAY_LAT not in df_columns or row[STAY_LAT] == -1 else 1
+
+    return pd.Series([duration, lat, long, stay_not])
+
+
+def _get_stays_by_lat_long_dur(df_by_user: pd.DataFrame) -> np.ndarray:
+    """
+    Groups together consecutive traces that have the same value for latitude, longitude, and duration.
+
+    Args:
+        df_by_user (pd.DataFrame): Dataframe filtered by user.
+
+    Returns:
+        np.ndarray: Array of stay group indices.
+    """
+    lat = df_by_user['lat'].to_numpy()
```

Comment:
do we need the `to_numpy()`? The [docs](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_numpy.html) state that they don't guarantee a copy is not made.

Reply:
We use 'lat' to do array operations later in the code. We use to_numpy as the loop-based operations are more efficient on numpy arrays than on pd.Series (if we had directly used df_by_user['lat'])

Also, the copying should not be a concern because we are not changing the 'lat' array. We are only reading the values for comparison.

Let me know if I missed something, or if you think otherwise.